### PR TITLE
[BR-5] Cover with tests method findById() + first level cache of Session

### PR DESCRIPTION
**Pull request by ivanbolshak**
- 94% coverage by line for Session class

### CODE REVIEW

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    private GenericJdbcDAO mockJdbcDAO;
+    private Map<EntityKey<?>, Object> entitiesCacheExpected = new HashMap<>();
+    private Map<EntityKey<?>, Object[]> entitiesSnapshotsExpected = new HashMap<>();
+
+    private Field entitiesCacheField;
+    private Field entitiesSnapshotsField;
+
+    private AtomicInteger personIdSequence = new AtomicInteger(0);
+
+    @SneakyThrows
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: we not used @SneakyThrows in test

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    @BeforeEach
+    public void initData() {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    @AfterEach
+    public void cleanData() {
+        entitiesCacheExpected.clear();
+        entitiesSnapshotsExpected.clear();
+    }
+
+    @Test
+    void shouldReturnPersonById_positive() {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        Assertions.assertEquals(0, entitiesCacheExpected.size());
+        Assertions.assertEquals(0, entitiesSnapshotsExpected.size());
+        Mockito.verify(mockJdbcDAO, times(0)).update(any());
+
+        System.out.println();
+    }
+
+    @Test
+    void shouldCloseSessionWithNullInEntityField() {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        Assertions.assertEquals(0, entitiesCacheExpected.size());
+        Assertions.assertEquals(0, entitiesSnapshotsExpected.size());
+
+        Mockito.verify(mockJdbcDAO, times(0)).update(any());
+    }
+
+    @Test
+    void shouldCloseSessionWithUpdateRecord() {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        Assertions.assertEquals(0, entitiesCacheExpected.size());
+        Assertions.assertEquals(0, entitiesSnapshotsExpected.size());
+
+        Mockito.verify(mockJdbcDAO, times(1)).update(any());
+    }
+
+    @Test
+    void shouldCloseSessionWithUpdateIfNullFieldRecord() {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        Assertions.assertEquals(0, entitiesCacheExpected.size());
+        Assertions.assertEquals(0, entitiesSnapshotsExpected.size());
+
+        Mockito.verify(mockJdbcDAO, times(1)).update(any());
+    }
+
+    @Test
+    void shouldCloseSessionWithUpdateIfTwoNullsFieldRecord() {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        PersonSessionTest person = new PersonSessionTest();
+        person.setId(personIdSequence.incrementAndGet());
+        person.setFirstName(RandomStringUtils.randomAlphabetic(3, 10));
+        person.setLastName(RandomStringUtils.randomAlphabetic(3, 10));
+        person.setAge(new Random().nextInt(0, 100));
+        person.setMale(RandomStringUtils.randomAlphabetic(1));
+        return person;
+    }
+
+    @SneakyThrows
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: we not used @SneakyThrows in test

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+    private void addPersonToTastableScope(PersonSessionTest person) {
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        addPersonToEntityExpected(person);
+        addEntityToSnapshots(getEntityKeyByPerson(person), person);
+
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java]: remove this line

File: src/test/java/org/svydovets/session/SessionTest.java [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        entitiesSnapshotsField.setAccessible(true);
+        entitiesSnapshotsField.set(sessionTestable, entitiesSnapshotsExpected);
+    }
+
+    @SneakyThrows
+    private Object getFieldValue(Object entity, Field fields) {
+        fields.setAccessible(true);
+        return fields.get(entity);
+    }
+}
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/session/SessionTest.java:231]: add last empty line

File: src/test/java/org/svydovets/baseEntity/PersonSessionTest.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    private String firstName;
+
+    @Column(name = "last_name")
+    private String lastName;
+
+    @Column
+    private Integer age;
+
+    private String male;
+}
```
Comment (AlexandrNavozenko) [inline src/test/java/org/svydovets/baseEntity/PersonSessionTest.java:31]: add last empty line

File: src/main/java/org/svydovets/session/Session.java [hunk 13]
```diff
@@ -57,7 +57,15 @@ private boolean hasChanged(Map.Entry<EntityKey<?>, Object> entry) {
         Field[] fields = ReflectionUtils.getEntityFieldsSortedByName(entityKey.clazz());
         Object[] snapshots = entitiesSnapshots.get(entityKey);
         for (int i = 0; i < snapshots.length; i++) {
-            if (!snapshots[i].equals(getFieldValue(entity, fields[i]))) {
+
```
Comment (Pan14ek) [inline src/main/java/org/svydovets/session/Session.java:60]: new comment

**Review by Pan14ek — COMMENTED** [review]
TEST

**Pull request merged**